src/test/run-pass/union/union-nonzero.rs

```diff
@@ -0,0 +1,50 @@
+// run-pass
+#![allow(dead_code)]
+
+// Tests that unions aren't subject to unsafe non-zero/niche-filling optimizations.
+//
+// For example, if a union `U` can contain both a `&T` and a `*const T`, there's definitely no
+// bit-value that an `Option<U>` could reuse as `None`; this test makes sure that isn't done.
+//
+// Secondly, this tests the status quo to not apply such optimizations to types containing unions
+// even if they're theoretically possible. (discussion: https://github.com/rust-lang/rust/issues/36394)
```

Comment:
I think this can be read as a guarantee we make about `repr(Rust)` unions in general and not `MaybeUninit<T>` specifically. Can you please disclaim such beliefs by extending the comment here?

Reply:
Big picture question: I wrote that assuming that comments in tests aren't taken as user-facing guarantees, but rather just internal explanations. Do you think that's a good assumption in general? Or should I write such stuff differently?

As for this specific comment, I used the term "status quo" to indicate that this is just how Rust works right now, but could change in the future; I figured it'd be useful to know if we change the behavior accidentally.

Maybe put in brackets after "status quo" the short statement "(not a guarantee; subject to change!)"?

Reply:
> Do you think that's a good assumption in general? Or should I write such stuff differently?

I think they can be taken as user facing guarantees; some will interpret them in that way and then it's best to disclaim it; it doesn't cost too much.

> Maybe put in brackets after "status quo" the short statement "(not a guarantee; subject to change!)"?

👍 That works well :)

Reply:
Thanks, I'll keep that in mind.

Fixed!